# Car on budget
The aim is to clean the data and analyze the included used car listings.
The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

In [1]:
import pandas as pd 
import numpy as np

In [2]:
autos=pd.read_csv("autos.csv",encoding="Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### The dataset contains 20 columns, most of which are strings.

 Some columns have null values, but none have more than ~20% null values.

The column names use camelcase instead of Python's preferred snakecase, which means we cannot just replace spaces with underscores, here we will use df.coloumns attribute to replace camelcase with snakcase

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We will change the columns from camelcase to Snakecase for e.g:
* yearOfRegistration to registration_year
* monthOfRegistration to registration_month
* notRepairedDamage to unrepaired_damage
* dateCreated to ad_created
   
  And for the rest of the column names.

In [7]:
arr=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_pS', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns=arr

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  $5,000  control          bus                  2004   
1  privat    Angebot  $8,500  control    limousine                  1997   
2  privat    Angebot  $8,990     test    limousine                  2009   
3  privat    Angebot  $4,350  control   kleinwagen                  2007   
4  privat    Angebot  $1,350     test        kombi                  2003   

     gearbox  power_pS   model   odometer  month_of_registration fuel_type  \
0    manuell       158  andere  150,000km                      3       lpg   
1  automatik       286     7er  150,000km                      6    benzin   
2    manuell       102    golf   70,000km                      7    benzin   
3  automatik        71  fortwo   70,000km                      6    benzin   
4    manuell         0   focus  150,000km                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [9]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  year_of_registration  gearbox      power_pS  model  \
count         44905          50000.000000    47320  50000.000000  47242   
unique            8                   NaN        2           NaN    245   
top       limousine                   NaN  manuell           NaN   golf   
freq          12859                   NaN    36993           NaN   4024   
mean            NaN           2005.073280      NaN    116.355920    NaN   
std             NaN            105.712813      NaN    209.216627    NaN   
min             NaN           1000.000000      NaN      0.000000    NaN   
25%             NaN           1999.000000      NaN     70.000000    NaN   
50%             NaN           2003.000000      NaN    105.000000    NaN   
75%             NaN           2008.000000      NaN    150.000000    NaN   
max             NaN           9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Notes:-
* Any columns that have mostly one value that are candidates to be dropped
* Any columns that need more investigation.
* Any examples of numeric data stored as text that needs to be cleaned.

From above we can see that nr_of_pictures can be drop , because their no use analyse such coloumn. 

The seller and offer_type has same kind of repeated values

In [10]:
autos["price"]=autos["price"].str.replace("$","").str.replace(",","").astype(float)

In [11]:
autos["odometer"]=autos["odometer"].str.replace("km","").str.replace(",","").astype(float)

In [12]:
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

In [13]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  5000.0  control          bus                  2004   
1  privat    Angebot  8500.0  control    limousine                  1997   
2  privat    Angebot  8990.0     test    limousine                  2009   
3  privat    Angebot  4350.0  control   kleinwagen                  2007   
4  privat    Angebot  1350.0     test        kombi                  2003   

     gearbox  power_pS   model  odometer_km  month_of_registration fuel_type  \
0    manuell       158  andere     150000.0                      3       lpg   
1  automatik       286     7er     150000.0                      6    benzin   
2    manuell       102    golf      70000.0                      7    benzin   
3  automatik        71  fortwo      70000.0                      6    benzin   
4    manuell         0   focus     150000.0                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [14]:
print(autos["odometer_km"].unique().shape)
autos["price"].unique().shape

(13,)


(2357,)

In [15]:
print(autos["odometer_km"].describe())
autos["price"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
print(autos["odometer_km"].value_counts(ascending=False))
autos["price"].value_counts(ascending=False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64


0.0        1421
500.0       781
1500.0      734
2500.0      643
1200.0      639
           ... 
6202.0        1
18310.0       1
898.0         1
11240.0       1
789.0         1
Name: price, Length: 2357, dtype: int64

In [17]:
print(autos[autos["price"].between(100,10000000)].shape)
autos=autos[autos["price"].between(100,10000000)]

(48231, 20)


I have filter the value in price column that has price less than 100$ because , their almost no car in that range of price 

In [18]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  5000.0  control          bus                  2004   
1  privat    Angebot  8500.0  control    limousine                  1997   
2  privat    Angebot  8990.0     test    limousine                  2009   
3  privat    Angebot  4350.0  control   kleinwagen                  2007   
4  privat    Angebot  1350.0     test        kombi                  2003   

     gearbox  power_pS   model  odometer_km  month_of_registration fuel_type  \
0    manuell       158  andere     150000.0                      3       lpg   
1  automatik       286     7er     150000.0                      6    benzin   
2    manuell       102    golf      70000.0                      7    benzin   
3  automatik        71  fortwo      70000.0                      6    benzin   
4    manuell         0   focus     150000.0                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [19]:
autos[['date_crawled','date_created','last_seen']][0:5]

date_crawled         date_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [20]:
autos["date_crawled"].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05 14:06:30    0.000021
2016-03-05 14:06:40    0.000021
2016-03-05 14:07:04    0.000021
2016-03-05 14:07:08    0.000021
2016-03-05 14:07:21    0.000021
                         ...   
2016-04-07 14:30:09    0.000021
2016-04-07 14:30:26    0.000021
2016-04-07 14:36:44    0.000021
2016-04-07 14:36:55    0.000021
2016-04-07 14:36:56    0.000021
Name: date_crawled, Length: 46578, dtype: float64

In [21]:
autos["date_created"].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11 00:00:00    0.000021
2015-08-10 00:00:00    0.000021
2015-09-09 00:00:00    0.000021
2015-11-10 00:00:00    0.000021
2015-12-05 00:00:00    0.000021
                         ...   
2016-04-03 00:00:00    0.038855
2016-04-04 00:00:00    0.036926
2016-04-05 00:00:00    0.011797
2016-04-06 00:00:00    0.003255
2016-04-07 00:00:00    0.001244
Name: date_created, Length: 76, dtype: float64

In [22]:
autos["last_seen"].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05 14:45:46    0.000021
2016-03-05 14:46:02    0.000021
2016-03-05 14:49:34    0.000021
2016-03-05 15:16:11    0.000021
2016-03-05 15:16:47    0.000021
                         ...   
2016-04-07 14:58:44    0.000062
2016-04-07 14:58:45    0.000021
2016-04-07 14:58:46    0.000021
2016-04-07 14:58:48    0.000062
2016-04-07 14:58:50    0.000062
Name: last_seen, Length: 38237, dtype: float64

Most of the cars were added in year 2015 and not get sold for 1 year 

In [23]:
autos["year_of_registration"].describe()

count    48231.000000
mean      2004.728867
std         87.891383
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

This registration year's min value is 1000 and max is 9999 which is the wrong data because at that time the car was invented i.e before 19th cetury and current year is 2020 so we cannot have the registration of cars of future because they are not made yet , so these row data is redundent.

So we would select the year the data range of registration year from 1900-2020

In [24]:
autos=autos[autos["year_of_registration"].between(1900,2020)]

In [25]:
autos["year_of_registration"].value_counts(normalize=True)

2000    0.064382
2005    0.060379
1999    0.059736
2004    0.055982
2003    0.055857
          ...   
1929    0.000021
1939    0.000021
1938    0.000021
1931    0.000021
1952    0.000021
Name: year_of_registration, Length: 81, dtype: float64

In [26]:
brand=autos["brand"].unique()

Now we will understand the price scenario of different brands , that how much average price of a particular brand is their. 
I am selecting the top 20 brands of cars which are more in number and then we use aggregation on those brands to their average price.

In [27]:
bd=autos["brand"].value_counts().head(20).index
bd

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [28]:
brand_price={}
for c in bd:
    mn=autos.loc[autos["brand"]==c,"price"]
    mn=mn.mean()
    brand_price[c]=mn
    
brand_price

{'volkswagen': 5559.968156587788,
 'bmw': 8543.978260869566,
 'opel': 2971.6082375478927,
 'mercedes_benz': 8570.76869865975,
 'audi': 9259.510248372317,
 'ford': 4053.7575215966635,
 'renault': 2450.9015611448394,
 'peugeot': 3086.930281690141,
 'fiat': 2815.635782747604,
 'seat': 4353.146929824561,
 'skoda': 6394.309677419355,
 'nissan': 4681.94046008119,
 'mazda': 4075.319293478261,
 'smart': 3538.344927536232,
 'citroen': 3777.854625550661,
 'toyota': 5148.0032733224225,
 'hyundai': 5416.23382045929,
 'sonstige_autos': 45941.48791208791,
 'volvo': 4911.680459770115,
 'mini': 10566.824940047962}

The sonstigue_autos is the most expensive brand that is to sold in the market and then on 2nd is mini . 

In [29]:
mean_mileage={}
for c in brand_price:
    row=autos.loc[autos["brand"]==c,"odometer_km"]
    row=row.mean()
    mean_mileage[c]=row

In [30]:
   mean_mileage

{'volkswagen': 129060.27850813128,
 'bmw': 132827.99389778794,
 'opel': 129512.45210727969,
 'mercedes_benz': 131079.76653696498,
 'audi': 129604.53339763684,
 'ford': 124361.03663985702,
 'renault': 128261.05810928014,
 'peugeot': 127316.9014084507,
 'fiat': 117408.14696485623,
 'seat': 122149.12280701754,
 'skoda': 111051.6129032258,
 'nissan': 118707.71312584574,
 'mazda': 124959.23913043478,
 'smart': 100833.33333333333,
 'citroen': 119985.31571218796,
 'toyota': 116219.31260229132,
 'hyundai': 107150.31315240083,
 'sonstige_autos': 90912.08791208791,
 'volvo': 138839.0804597701,
 'mini': 89100.71942446043}

In [31]:
new=pd.Series(mean_mileage,index=mean_mileage.keys())

In [32]:
new_p=pd.Series(brand_price,index=mean_mileage.keys())

In [33]:
df=pd.DataFrame(new_p,columns=['mean_price'])

In [34]:
df

mean_price
volkswagen       5559.968157
bmw              8543.978261
opel             2971.608238
mercedes_benz    8570.768699
audi             9259.510248
ford             4053.757522
renault          2450.901561
peugeot          3086.930282
fiat             2815.635783
seat             4353.146930
skoda            6394.309677
nissan           4681.940460
mazda            4075.319293
smart            3538.344928
citroen          3777.854626
toyota           5148.003273
hyundai          5416.233820
sonstige_autos  45941.487912
volvo            4911.680460
mini            10566.824940

In [35]:
df["mileage"]=new

In [36]:
df

mean_price        mileage
volkswagen       5559.968157  129060.278508
bmw              8543.978261  132827.993898
opel             2971.608238  129512.452107
mercedes_benz    8570.768699  131079.766537
audi             9259.510248  129604.533398
ford             4053.757522  124361.036640
renault          2450.901561  128261.058109
peugeot          3086.930282  127316.901408
fiat             2815.635783  117408.146965
seat             4353.146930  122149.122807
skoda            6394.309677  111051.612903
nissan           4681.940460  118707.713126
mazda            4075.319293  124959.239130
smart            3538.344928  100833.333333
citroen          3777.854626  119985.315712
toyota           5148.003273  116219.312602
hyundai          5416.233820  107150.313152
sonstige_autos  45941.487912   90912.087912
volvo            4911.680460  138839.080460
mini            10566.824940   89100.719424

BMW, Mercedes and mini are the kings of the German auto ebay market. They have the 3 most popular and three most expensive car brands on average. With adui as the fourth most popular it is clear that Germans like to buy German.

Mileage has a significant affect on the price of vehicles. The more miles on the clock, the less expensive the car will tend to be.